File: java/Observer.java
import java.util.ArrayList;
import java.util.List;

interface Observateur {
    void notifier(String evenement);
}

class ObservateurEmail implements Observateur {
    @Override
    public void notifier(String evenement) {
        System.out.println("[Email] Nouvel événement reçu : " + evenement);
    }
}

class ObservateurLog implements Observateur {
    @Override
    public void notifier(String evenement) {
        System.out.println("[Log] Événement enregistré : " + evenement);
    }
}

class Sujet {
    private List<Observateur> observateurs = new ArrayList<>();

    public void ajouterObservateur(Observateur observateur) {
        observateurs.add(observateur);
    }

    public void notifierTous(String evenement) {
        for (Observateur o : observateurs) {
            o.notifier(evenement);
        }
    }
}

public class Observer {
    public static void main(String[] args) {
        Sujet sujet = new Sujet();
        sujet.ajouterObservateur(new ObservateurEmail());
        sujet.ajouterObservateur(new ObservateurLog());

        sujet.notifierTous("Nouvelle commande passée");
    }
}

File: python/observer.py
class Observateur:
    def notifier(self, evenement):
        raise NotImplementedError


class ObservateurEmail(Observateur):
    def notifier(self, evenement):
        print(f"[Email] Nouvel événement reçu : {evenement}")


class ObservateurLog(Observateur):
    def notifier(self, evenement):
        print(f"[Log] Événement enregistré : {evenement}")


class Sujet:
    def __init__(self):
        self.observateurs = []

    def ajouter_observateur(self, observateur):
        self.observateurs.append(observateur)

    def notifier_tous(self, evenement):
        for observateur in self.observateurs:
            observateur.notifier(evenement)


if __name__ == "__main__":
    sujet = Sujet()
    sujet.ajouter_observateur(ObservateurEmail())
    sujet.ajouter_observateur(ObservateurLog())

    sujet.notifier_tous("Nouvelle commande passée")